src/snowflake/snowpark/dataframe.py

```diff
@@ -878,7 +879,9 @@ def __copy__(self) -> "DataFrame":
             new_plan._query_params = self._select_statement.query_params
         else:
             new_plan = copy.copy(self._plan)
-        return DataFrame(self._session, new_plan)
+        df = DataFrame(self._session, new_plan)
+        df._ast_id = self._ast_id
```

Comment:
If `DataFrame` never, ever changes in place, this approach is safe. If any meaningful `DataFrame` API mutates the object in place, using the same variable ID for two objects will lead to very surprising results at some point in the future.

We might have this pattern in other places, too.

Reply:
What is reason for this, performance?

Reply:
I think this a copy-by-reference vs. deep-copy concern here. Indeed we should be careful regarding this. Atm, this is done using copy-by-reference. However, we may want to add a separate IR API to emulate the copy behavior (maybe reuse cache?)

cc @sfc-gh-azwiegincew 

I'll add a JIRA with todo so we can pickup the copy behavior there.